File: 2020/02_PasswordPhilosophy/policy.py
# ======================================================================
# Password Philosophy
#   Advent of Code 2020 Day 02 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         p o l i c y . p y
# ======================================================================
"A password policy for the Advent of Code 2020 Day 02 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import re
import collections

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
re_text = re.compile('([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)')

# ======================================================================
#                                                                 Policy
# ======================================================================


class Policy(object):   # pylint: disable=R0902, R0205
    "A password policy object for Password Philosophy"

    def __init__(self, text=None):

        # 1. Set the initial values
        self.text = text
        self.low = 0
        self.high = 0
        self.letter = None
        self.password = None

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            result = re_text.fullmatch(text)
            if result is None:
                print("unable to match [%s]" % text)
            else:
                self.low = int(result.group(1))
                self.high = int(result.group(2))
                self.letter = result.group(3)
                self.password = result.group(4)

    def count(self):
        "Returns the number of times the selected letter appears"
        # 1. Count all the letters in the policy
        assert self.password is not None
        letters = collections.Counter(self.password)

        # 2. Check the number of occurences of the selected letter
        return letters[self.letter]

    def locations(self):
        "Returns the locations of the selected letter"

        # 1. Start with nothing
        result = set()

        # 2. Loop for all of the letters in the password
        for index, letter in enumerate(self.password):

            # 3. If this is the letter we want, add the locations
            if self.letter == letter:
                result.add(1 + index)

        # 4. Return the locations
        return result

    def is_valid(self):
        "Returns True if the password is valid according to the policy"

        # 1. Get the number of times the letter appears
        count = self.count()

        # 2. Check the number of occurences of the selected letter
        return count >= self.low and count <= self.high

    def is_valid2(self):
        "Returns True if the password is valid according to the policy"

        # 1. Get the number of times the letter appears
        locs = self.locations()

        # 2. Count the number of occurences of the selected letter
        result = 0
        if self.low in locs:
            result += 1
        if self.high in locs:
            result += 1

        # 3. There can be only one
        return result == 1


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        p o l i c y . p y                       end
# ======================================================================

File: 2020/02_PasswordPhilosophy/policy.lua
-- ======================================================================
-- Password Philosophy
--   Advent of Code 2020 Day 02 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         p o l i c y . l u a
-- ======================================================================
-- "Policy for the Advent of Code 2020 Day 02 puzzle"

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Policy = {}
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
CAPTURE = "(%d+)-(%d+) (%a): (%a+)"

-- ======================================================================
--                                                                 Policy
-- ======================================================================


function Policy:Policy (o)
  -- "Object for Password Philosophy"


  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or ''
  o.low = 0
  o.high = 0
  o.letter = '?'
  o.password = '?'

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Policy:_process_text(text)
  -- "Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)
  
  -- 1. Break down the policy
  low, high, letter, password = string.match(text, CAPTURE)
  
  -- 2. Save the parts
  self.text = text
  self.low = tonumber(low)
  self.high = tonumber(high)
  self.letter = letter
  self.password = password

end

function Policy:check_password()
  -- Check that the password is valid per the policy in effect
  
  if self.part2 then
    return self:check_password_two()
  end  
  return self:check_password_one()
  
end

function Policy:check_password_one()
  -- Check that the password is valid per the policy in part one
  
  -- 1. Get count of the specified character
  count = select(2, string.gsub(self.password, self.letter, ""))
  
  -- 2. Return true if the count if the character in the password is valid
  return count >= self.low and count <= self.high
  
end

function Policy:check_password_two()
  -- Check that the password is valid per the policy in part two
  
  -- 1. Get count of the specified character at the specified locations
  count = 0
  if self.letter == self.password:sub(self.low, self.low) then
    count = count + 1
  end  
  if self.letter == self.password:sub(self.high, self.high) then
    count = count + 1
  end  
  
  -- 2. Return true if the count if the character in the password is valid
  return count == 1
  
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
return Policy

-- ======================================================================
-- end                       p o l i c y . l u a                      end
-- ======================================================================
